graylog2-server/src/main/java/org/graylog2/cluster/leader/AutomaticLeaderElectionService.java

```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        super.startUp();
+        executionThread = Thread.currentThread();
+    }
+
+    @Override
+    protected void triggerShutdown() {
+        executionThread.interrupt();
+    }
+
+    @Override
+    public boolean isLeader() {
+        return isLeader;
+    }
+
+    @Override
+    protected void run() throws Exception {
+        while (isRunning()) {
+            try {
+                runIteration();
+            } catch (Throwable t) {
+                log.error("Exception while acquiring/renewing leader lock.", t);
+            }
+        }
+    }
+
+    private void runIteration() {
+        final boolean wasLeader = isLeader;
+
+        try {
+            isLeader = lockService.lock(RESOURCE_NAME).isPresent();
+            lastSuccess = System.nanoTime();
+        } catch (Exception e) {
+            log.error("Unable to acquire/renew leader lock.", e);
+
+            final Duration timeSinceLastSuccess = Duration.ofNanos(System.nanoTime() - lastSuccess);
+            if (wasLeader && timeSinceLastSuccess.compareTo(lockTTL) >= 0) {
+                log.error("Failed for {} to renew leader lock. Forcing fallback to follower role.",
+                        timeSinceLastSuccess);
+                isLeader = false;
+            }
+        }
+
+        if (wasLeader != isLeader) {
+            handleLeaderChange(isLeader);
+        }
+
+        pauseBeforeNextIteration(wasLeader && !isLeader);
+    }
+
+    private void pauseBeforeNextIteration(boolean isDowngrade) {
+        try {
+            if (isDowngrade) {
+                log.info("Pausing leader-lock acquisition attempts for {} after downgrade from leader.", lockTTL);
+                Thread.sleep(lockTTL.toMillis());
+                log.info("Resuming leader-lock acquisition attempts every {}.", pollingInterval);
+            } else {
+                Thread.sleep(pollingInterval.toMillis());
+            }
+        } catch (InterruptedException e) {
+            // OK, we are shutting down.
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    private void handleLeaderChange(boolean isLeader) {
+        if (isLeader) {
+            log.info("Leader changed. This node is now the leader.");
+        } else {
+            log.error("Leader changed. This node lost the leader role. This should not happen. The node will " +
+                    "attempt to gracefully transition to assuming a follower role.");
+        }
+
+        // Ensure the nodes collection is up to date before we publish the event
+        try {
+            nodePingThread.doRun();
+        } catch (Exception e) {
+            log.error("Unable to trigger update of nodes collection.");
+        }
+        eventBus.post(new LeaderChangedEvent());
+    }
+
+    @Override
+    public void giveUpLeader() {
+        lockService.unlock(RESOURCE_NAME).ifPresent(l -> log.info("Gave up leader lock on shutdown"));
```

Comment:
I'm looking at that

Reply:
I think this is better because right now, the service is still running when we call it :grimacing: 

```
--- graylog2-server/src/main/java/org/graylog2/cluster/leader/AutomaticLeaderElectionService.java
+++ graylog2-server/src/main/java/org/graylog2/cluster/leader/AutomaticLeaderElectionService.java
@@ -142,6 +142,7 @@ public class AutomaticLeaderElectionService extends AbstractExecutionThreadServi
 
     @Override
     public void giveUpLeader() {
+        stopAsync().awaitTerminated();
         isLeader = false;
         lockService.unlock(RESOURCE_NAME).ifPresent(l -> log.info("Gave up leader lock on shutdown"));
     }
diff --git graylog2-server/src/main/java/org/graylog2/cluster/leader/LeaderElectionService.java graylog2-server/src/main/java/org/graylog2/cluster/leader/LeaderElectionService.java
index 46342d574e..3ffe5cc61f 100644
--- graylog2-server/src/main/java/org/graylog2/cluster/leader/LeaderElectionService.java
+++ graylog2-server/src/main/java/org/graylog2/cluster/leader/LeaderElectionService.java
@@ -29,6 +29,7 @@ public interface LeaderElectionService {
 
     /**
      * Releases leadership claim of the current node.
+     * This will also shut down the LeaderElectionService.
      */
     void giveUpLeader();
 }
```

Reply:
see https://github.com/Graylog2/graylog2-server/pull/11719